### PR TITLE
[TRTLLM-13035][feat] Native /v1/embeddings dynamic batching for encoder-only models

### PR DESCRIPTION
**Pull request by xwang233**
## Release Notes

* **New Features**
  * Introduced OpenAI-compatible `/v1/embeddings` endpoint for encoder-only models (BERT-style classifiers, text-embedding models, reward models) via new `trtllm-serve embeddings` command
  * Added native dynamic batching to coalesce concurrent embedding requests for improved throughput
  * Added support for Qwen3-Embedding model architecture

* **Documentation**
  * Added comprehensive embeddings feature documentation with quick-start guide and API specification
  * Updated supported models list to include text-embedding architectures

## Description

This PR adds a native, OpenAI-compatible **`POST /v1/embeddings`** endpoint to `trtllm-serve`, with **dynamic batching** built in — so customers serving encoder-only / embedding models no longer need NVIDIA Triton Inference Server's dynamic batcher in front of TRT-LLM. (Ref: TRTLLM-13035)

**How it works**

- A new **`trtllm-serve embeddings <model>`** subcommand launches an `encode_only=True` LLM and exposes `/v1/embeddings` (plus `/health`, `/version`, `/v1/models`).
- A standalone **lightweight async dynamic batcher** (`tensorrt_llm/serve/encode_batcher.py`) sits in front of the existing synchronous `llm.encode()` fast path. A single background worker coalesces independent concurrent requests and flushes on any of three triggers — batch reaches `max_batch_size`, the next item would exceed `max_num_tokens`, or the `max_queue_delay` hold window elapses — mirroring Triton's `preferred_batch_size` / `max_queue_delay_microseconds` / `max_queue_size` knobs. The forward runs via `run_in_executor` so the event loop stays live; backpressure returns 429, oversized inputs 400.
- We deliberately **do not** re-enter the PyExecutor in-flight-batching loop: encoder models run a single short forward with no KV cache / sampler / scheduler, so the per-step IFB host overhead would dominate. The batcher calls `EncoderExecutor.batch_forward()` (single forward, `kv_cache_manager=None`, CUDA-graph capable) unchanged.

**Models supported (model-output-agnostic serialization)**

The endpoint serializes whatever per-request tensor `encode()` emits into the OpenAI schema:

- **BERT sequence classifiers** — `[num_labels]` class-logit vector.
- **Decoder reward models** (e.g. Qwen2.5-Math-PRM-7B) — per-token `[seq_len, num_labels]`.
- **Qwen3-Embedding family** (0.6B / 4B / 8B) — a new **`Qwen3ForTextEmbedding`** model (reuses the Qwen3 decoder backbone, drops `lm_head`, returns the **L2-normalized last-token hidden state**). The launch path auto-remaps `Qwen3ForCausalLM → Qwen3ForTextEmbedding` via a `model_kwargs` architecture override, so no special flags are needed. This makes the endpoint useful for real sentence-embedding workloads, not just classifier/reward logits.

Here "encoder-only" means the **encode execution mode** (one forward, no autoregression), not the encoder *architecture* — the PRM and Qwen3 models are causal decoders run in encode mode.

**Scope**

- **Single-GPU per server** by design. Scale out with data-parallel replicas behind a load balancer (documented in `embeddings.md`); this matches how vLLM / SGLang / Triton scale embeddings. The subcommand does not expose TP/PP/CP flags (a `--config` setting >1 fails fast with a clear message).
- `encoding_format` (`float` / `base64`) is supported. `dimensions` (Matryoshka) is rejected with 400 for the current non-Matryoshka models, matching vLLM.

**API change:** additive only — new `EmbeddingRequest` / `EmbeddingResponseData` / `EmbeddingResponse` / `EmbeddingUsageInfo` in `openai_protocol.py` and an updated `tests/unittest/api_stability/references/llm.yaml`. Backward compatible → please apply the **`api-compatible`** label.

## Test Coverage

- `tests/unittest/llmapi/apps/test_encode_batcher.py` — batcher unit tests (CPU, no GPU): single submission, concurrent coalescing, count/token-budget/hold-window flush triggers, 429 on full queue, 400 on oversized input, per-batch failure isolation, event-loop responsiveness.
- `tests/unittest/llmapi/test_embedding_arch_routing.py` — `Qwen3ForCausalLM → Qwen3ForTextEmbedding` architecture-override resolution (CPU).
- `tests/unittest/llmapi/apps/_test_openai_embeddings.py` — serve-level `/v1/embeddings` integration (GPU): BERT classifier, PRM-7B reward model, and Qwen3-Embedding `0.6B`/`8B` (dim + unit-norm + batch indexing + concurrent coalescing). Registered in `tests/integration/test_lists/test-db/l0_l40s.yml`.
- `tests/integration/defs/accuracy/test_llm_api_pytorch_encode.py` — Qwen3-Embedding `encode()` vs a HuggingFace `AutoModel` last-token + L2-norm reference (`rtol/atol=1.5e-2`), parametrized over `0.6B`/`8B`. Registered in `l0_l40s.yml`.

GPU-validated on single L40: batcher units + all serve-level + accuracy tests pass.

## PR Checklist

- [x] PR description clearly explains what and why.
- [x] PR follows TRT-LLM coding guidelines.
- [x] Test cases are provided for new code paths (batcher units, arch-routing unit, serve-level integration, accuracy vs HF).
- [x] API change is additive; `api-compatible` label requested (touches `api_stability/references/llm.yaml`).
- [x] No new dependencies introduced.
- [x] Documentation updated (`docs/source/features/embeddings.md`, `supported-models.md`).
- [x] Reviewers assigned are appropriate for the serve / model code paths.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

**Comment by xwang233** [comment]
/bot run --stage-list "A10-PyTorch-1, A10-PyTorch-2, L40S-PyTorch-1, L40S-PyTorch-2" --disable-fail-fast

File: tests/integration/test_lists/test-db/l0_l40s.yml [hunk 1]
```diff
@@ -27,6 +27,11 @@ l0_l40s:
   - unittest/_torch/modeling/test_modeling_step3p7vl.py
   - test_e2e.py::test_ptp_scaffolding[DeepSeek-R1-Distill-Qwen-7B-DeepSeek-R1/DeepSeek-R1-Distill-Qwen-7B]
   - unittest/llmapi/apps/_test_openai_chat_multimodal.py::test_single_chat_session_image_embeds -m needs_l40s
+  # Encoder-only /v1/embeddings dynamic batching (BERT classifier + PRM-7B reward + Qwen3-Embedding)
+  - unittest/llmapi/apps/_test_openai_embeddings.py
+  # Qwen3-Embedding text-embedding accuracy vs HuggingFace (0.6B small/fast + 8B large)
+  - accuracy/test_llm_api_pytorch_encode.py::TestEncoderEncode::test_qwen3_text_embedding_matches_huggingface[qwen3-embedding-0.6b]
+  - accuracy/test_llm_api_pytorch_encode.py::TestEncoderEncode::test_qwen3_text_embedding_matches_huggingface[qwen3-embedding-8b]
```
Comment (jieli-matrix) [inline tests/integration/test_lists/test-db/l0_l40s.yml:35]: you may consider add e2e test into qa test list if need more machine coverage, thx!

**Review by jieli-matrix — APPROVED** [review]

File: tests/integration/test_lists/test-db/l0_l40s.yml (hunk 1 above)
Comment (xwang233) [reply tests/integration/test_lists/test-db/l0_l40s.yml:35]: thanks! I wasn't aware of those qa test coverage and I just added the entries.

**Comment by xwang233** [comment]
/bot run --stage-list "A10-PyTorch-1, A10-PyTorch-2, L40S-PyTorch-1, L40S-PyTorch-2" --disable-fail-fast

File: tensorrt_llm/_torch/models/modeling_qwen3.py [hunk 2]
```diff
@@ -299,3 +300,67 @@ def __init__(
             model_config._frozen = False
             model_config.mapping = self.mapping_with_cp
             model_config._frozen = True
+
+
+@register_auto_model("Qwen3ForTextEmbedding")
+class Qwen3ForTextEmbedding(DecoderModelForCausalLM[Qwen3Model, Qwen3Config]):
+    """Qwen3-Embedding family (0.6B / 4B / 8B).
+
+    The Qwen3 decoder backbone exposed as a sentence-embedding model: ``forward``
```
Comment (2ez4bz) [inline tensorrt_llm/_torch/models/modeling_qwen3.py]: Nit: single backticks on changed lines (our python docs don't use / support this anyway).

File: tensorrt_llm/_torch/models/modeling_qwen3.py [hunk 3]
```diff
@@ -299,3 +300,67 @@ def __init__(
...
+    returns the L2-normalized last-token hidden state (the sentence embedding)
+    instead of vocabulary logits. This matches the model's sentence-transformers
+    pipeline (Transformer -> Pooling[last-token] -> Normalize) and SGLang's
+    ``Pooler(PoolingType.LAST, normalize=True)``.
+
+    ``lm_head`` is intentionally not allocated: like ``Qwen2ForRewardModel`` we call
```
Comment (2ez4bz) [inline tensorrt_llm/_torch/models/modeling_qwen3.py]: Nit: this makes more sense to have as a code comment above line 322 instead of in a public docstring.

**Review by 2ez4bz — APPROVED** [review]
model changes LGTM.

**Review by StanleySun639 — APPROVED** [review]

**Review by Superjomn — APPROVED** [review]
LGTM on the llmapi changes

File: tensorrt_llm/serve/encode_batcher.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+                self._max_num_tokens is not None
+                and num_tokens + len(request.token_ids) > self._max_num_tokens
+            ):
+                carry_over = request  # keeps the current batch within budget
+                break
+            batch.append(request)
+            num_tokens += len(request.token_ids)
+        return batch, carry_over
+
+    async def _dispatch(self, loop, batch: List[_QueuedRequest]) -> None:
```
Comment (QiJune) [inline tensorrt_llm/serve/encode_batcher.py:167]: EncodeBatcher._dispatch() unconditionally calls request.future.set_result(result). If a request task is cancelled while queued, its future is cancelled too. Dispatch then raises asyncio.InvalidStateError, terminates the sole batcher worker, and leaves all subsequent embedding requests hanging indefinitely.

Please check future.done() before calling set_result() and add a regression test covering request cancellation.

File: tensorrt_llm/serve/openai_server.py
Comment (QiJune) [inline tensorrt_llm/serve/openai_server.py:734]: PyTorchLLM._check_health() only checks whether _encoder_executor exists; it does not check whether embedding_batcher._worker is still alive. If the batcher worker exits, /health still returns 200 while all /v1/embeddings requests hang.

File: tensorrt_llm/serve/encode_batcher.py (hunk 4 above)
Comment (xwang233) [reply tensorrt_llm/serve/encode_batcher.py:167]: Guarded with `if not request.future.done()` . Added test `test_cancelled_request_does_not_kill_worker`

File: tensorrt_llm/serve/openai_server.py
Comment (xwang233) [reply tensorrt_llm/serve/openai_server.py:734]: added `EncodeBatcher.is_alive()` in health check

**Review by QiJune — APPROVED** [review]
LGTM
